Developer quality of life improvements (containerization, vendoring, GitHub CI testing)

welps opened the pull request:
It can be difficult to contribute to this project due to the underlying C/C++ libraries used by dependencies. The documentation helps, but it's easy to omit libraries that may come standard on some distros (I use arch btw) which can cause frustration for developers when they are trying to get to a working build.

This pull request provides the following enhancements:
- Provide working Dockerfiles for:
    - the target build (armf)
    - development builds
- GitHub Actions to prevent breakages by:
    - building the armf docker image and plato using `./build.sh slow`
    - building the development docker image and running tests
    - note that the time on these builds can be drastically improved by refactoring the steps that add packages from debian into an image that is hosted in a container registry
    - examples: [working armf run](https://github.com/welps/plato/runs/320256158) | [working dev run](https://github.com/welps/plato/runs/320256124) 
- Vendors the `thirdparty/` dependencies
    - This arose because freetype is hosted on savannah.nongnu.org and became inaccessible this weekend due to a disk array failure (see https://hostux.social/@fsfstatus)
- Removes need for developers to maintain their own `.cargo/config` as it is possible to maintain project-specific configuration as documented in https://doc.rust-lang.org/cargo/reference/config.html

I have tested the plato binary on my Kobo Forma that was created in the armf container and did not observe any issues. Note that instantiating the image into a container will just run `./build.sh slow` and if you want to mount the output to your host volume, you can run `docker run -v $(pwd)/target:/plato/target [nameofthecreatedimage]`

Let me know if there are any questions. Thanks so much for your work @baskerville , I really enjoy using plato.

baskerville commented:
> It can be difficult to contribute to this project due to the underlying C/C++ libraries used by dependencies. The documentation helps, but it's easy to omit libraries that may come standard on some distros (I use arch btw) which can cause frustration for developers when they are trying to get to a working build.

Contributors are supposed to build the ARM binary with `./build.sh` (in which case, the pre-built libraries are used).

> This pull request provides the following enhancements:
> 
> * Provide working Dockerfiles for:
>   
>   * the target build (armf)
>   * development builds

I like being able to run the compiler toolchain on my platform instead of having to run it within a virtual machine.

> * GitHub Actions to prevent breakages by:
>   
>   * building the armf docker image and plato using `./build.sh slow`

I'm afraid `./build.sh slow` isn't capable of skipping the dependencies that don't need to be rebuilt.

> * Vendors the `thirdparty/` dependencies
>   
>   * This arose because freetype is hosted on savannah.nongnu.org and became inaccessible this weekend due to a disk array failure (see https://hostux.social/@fsfstatus)

I don't feel comfortable vendoring the third party dependencies. If the breakage of *savannah.nongnu.org* happens regularly, we might just switch to using the *sourceforge* link.

> * Removes need for developers to maintain their own `.cargo/config` as it is possible to maintain project-specific configuration as documented in https://doc.rust-lang.org/cargo/reference/config.html

Nice.

welps commented:
I've rebased my branch off your recent master and made changes to address your comments. 

> Contributors are supposed to build the ARM binary with `./build.sh` (in which case, the pre-built libraries are used).

> I like being able to run the compiler toolchain on my platform instead of having to run it within a virtual machine.

These two comments are related to why I provided dockerfiles. You can continue to run the compiler toolchain on your platform, but I wanted to contribute to this project and went on a package safari (even with the documentation) that took a non-trivial amount of time to make sure that I could:
- build a working ARM binary from ./build.sh (fast and slow)
- build the emulator/importer on my host machine

You don't have to utilize this addition, but this
- makes it easier for others to contribute
- provides opportunity to automate builds and run tests in a controlled environment
     - with the additional benefit of saving you time as you do not need to look/deal with PRs where tests break or the ARM binary cannot compile due to developer error

> I don't feel comfortable vendoring the third party dependencies. If the breakage of _savannah.nongnu.org_ happens regularly, we might just switch to using the _sourceforge_ link.

I honestly can't keep up with whether sourceforge is spyware or not at the time. I reverted this change though since the `slow` option shouldn't be default and it's not necessary to deal with this.

welps commented:
@baskerville Thoughts on my responses?

baskerville commented:
Regarding the *armhf* workflow: I think the only required dependency is `pkg-config`. `patchelf` is only used by `dist.sh`. You need to use the following compiler toolchain: [gcc-linaro-4.9.4-2017.01-x86_64_arm-linux-gnueabihf.tar.xz](https://github.com/kobolabs/Kobo-Reader/blob/master/toolchain/gcc-linaro-4.9.4-2017.01-x86_64_arm-linux-gnueabihf.tar.xz) and symlink the relevant binaries: `find /path/to/the/extracted/archive/bin -name 'arm-*' -exec ln -s {} /usr/bin \;`.

Regarding the *dev* workflow: `cargo build --all-features` also builds the default binary: this is unnecessary. The set of dependencies also isn't minimal, we don't need `automake`, `ragel`, `cmake`, `jq`, `patchelf`, `g{cc,++}-arm-linux-gnueabihf`, or `libstdc++6`. Why is `libcairo2-dev` needed?

welps commented:
@baskerville 

I've removed the dependencies you've cited in the `armf` and `dev` flow. Both builds pass in GitHub CI ([armf](https://github.com/welps/plato/runs/378153357) | [dev](https://github.com/welps/plato/runs/378153378)). 

`jq` remains as it is required by `download.sh`. I believe these extraneous dependencies (like `libcairo2-dev`) were as a result of trying to support the `slow` build option.

For the `dev` flow, I agree that building the default binary is unnecessary, though `all-features` is used as a catch-all to prevent breakages in future PRs or changes.

baskerville commented:
Thanks.
